File: DSA/src/Questions/StackQueue/BasicCalculatorIII.java
package Questions.StackQueue;

import java.util.Stack;

/*
https://leetcode.com/problems/basic-calculator-iii/description/
Implement a basic calculator to evaluate a simple expression string.

The expression string contains only non-negative integers, '+', '-', '*', '/' operators, and open '(' and closing parentheses ')'. The integer division should truncate toward zero.

You may assume that the given expression is always valid. All intermediate results will be in the range of [-231, 231 - 1].

Note: You are not allowed to use any built-in function which evaluates strings as mathematical expressions, such as eval().

 

Example 1:

Input: s = "1+1"
Output: 2
Example 2:

Input: s = "6-4/2"
Output: 4
Example 3:

Input: s = "2*(5+5*2)/3+(6/2+8)"
Output: 21
 

Constraints:

1 <= s <= 104
s consists of digits, '+', '-', '*', '/', '(', and ')'.
s is a valid expression.

Explanation:
Immediate Result Update: Instead of storing results in a stack, we immediately add or subtract the current number to the result based on the last sign (+ or -).

Multiplication and Division: These operations are handled as soon as they are encountered. We update currentNumber directly with the product or quotient.

Parentheses:

When encountering (, push the current result and sign to the stack and reset them for the sub-expression.
When encountering ), pop from the stack to restore the previous context and continue.
Final Result: After processing the entire string, any remaining currentNumber is added to the result.

Complexity:
Time Complexity: O(n) — each character in the string is processed exactly once.
Space Complexity: O(1) in terms of the calculation, but O(m) if counting the space used by the stack for handling parentheses (where m is the depth of nested parentheses).
 */
public class BasicCalculatorIII {
    public boolean isDigit(char currentChar) {
        return '0' <= currentChar && currentChar <= '9';
    }

    public int calculate(String s) {
        Stack<Integer> stack = new Stack<>();
        int currentNumber = 0;
        char operation = '+';
        int n = s.length();

        for (int i = 0; i < n; i++) {
            char currentChar = s.charAt(i);

            if (isDigit(currentChar)) {
                currentNumber = currentNumber * 10 + (currentChar - '0');
            }

            if (currentChar == '(') {
                // Find the matching ')' and evaluate the expression inside
                int j = i, count = 0;
                for (; i < n; i++) {
                    if (s.charAt(i) == '(')
                        count++;
                    if (s.charAt(i) == ')')
                        count--;
                    if (count == 0)
                        break;
                }
                currentNumber = calculate(s.substring(j + 1, i));
            }

            if (!isDigit(currentChar) && currentChar != ' ' || i == n - 1) {
                switch (operation) {
                    case '+':
                        stack.push(currentNumber);
                        break;
                    case '-':
                        stack.push(-currentNumber);
                        break;
                    case '*':
                        stack.push(stack.pop() * currentNumber);
                        break;
                    case '/':
                        stack.push(stack.pop() / currentNumber);
                        break;
                }
                operation = currentChar;
                currentNumber = 0;
            }
        }

        // Sum up the values in the stack
        int result = 0;
        while (!stack.isEmpty()) {
            result += stack.pop();
        }

        return result;
    }
    public static void main(String[] args) {
        BasicCalculatorIII basicCalculatorIII = new BasicCalculatorIII();
        System.out.println(basicCalculatorIII.calculate("1+1")); // Output: 2
        System.out.println(basicCalculatorIII.calculate("6-4/2")); // Output: 4
        System.out.println(basicCalculatorIII.calculate("2*(5+5*2)/3+(6/2+8)")); // Output: 21
    }
}

File: DSA/Python/Questions/StackQueue/BasicCalculatorIII.py
"""
# Question: Basic Calculator III
# Link: https://leetcode.com/problems/basic-calculator-iii/

# Evaluate complex mathematical expressions with parentheses and operators

# Time Complexity: O(n)
# Space Complexity: O(n)

# Algorithm:
# 1. Process each character in expression
# 2. Handle numbers, operators (+, -, *, /), and parentheses
# 3. Use stack for operation tracking
# 4. Handle nested expressions recursively

# Key Components:
# - calculate(): Main expression evaluator
# - is_digit(): Helper for digit validation
# - Stack-based operation processing
"""


class BasicCalculatorIII:
    def is_digit(self, char: str) -> bool:
        return "0" <= char <= "9"

    def calculate(self, s: str) -> int:
        stack = []
        current_number = 0
        operation = "+"

        i = 0
        while i < len(s):
            char = s[i]

            if self.is_digit(char):
                current_number = current_number * 10 + int(char)

            if char == "(":
                # Find matching closing parenthesis
                count = 1
                j = i + 1
                while count != 0:
                    if s[j] == "(":
                        count += 1
                    if s[j] == ")":
                        count -= 1
                    j += 1
                current_number = self.calculate(s[i + 1 : j - 1])
                i = j - 1

            if (not self.is_digit(char) and char != " ") or i == len(s) - 1:
                if operation == "+":
                    stack.append(current_number)
                elif operation == "-":
                    stack.append(-current_number)
                elif operation == "*":
                    stack.append(stack.pop() * current_number)
                elif operation == "/":
                    stack.append(int(stack.pop() / current_number))

                operation = char
                current_number = 0

            i += 1

        return sum(stack)


def main():
    calculator = BasicCalculatorIII()
    print(calculator.calculate("1+1"))  # Output: 2
    print(calculator.calculate("6-4/2"))  # Output: 4
    print(calculator.calculate("2*(5+5*2)/3+(6/2+8)"))  # Output: 21


if __name__ == "__main__":
    main()
